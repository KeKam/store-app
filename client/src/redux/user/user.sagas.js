import { takeLatest, put, call, all } from 'redux-saga/effects';

import UserActionTypes from './user.types';
import {
  googleProvider,
  auth,
  createUserProfileDocument,
  getCurrentUser
} from '../../firebase/firebase';
import {
  signInSuccess,
  signInFailure,
  signOutSuccess,
  signOutFailure,
  signUpSuccess,
  signUpFailure
} from './user.actions';

export function* getSnapshotFromUserAuth(userAuth, additionalData) {
  try {
    const userRef = yield call(
      createUserProfileDocument,
      userAuth,
      additionalData
    );
    const userSnapshot = yield userRef.get();
    yield put(signInSuccess({ id: userSnapshot.id, ...userSnapshot.data() }));
  } catch (error) {
    yield put(signInFailure(error));
  }
}

export function* googleSignIn() {
  try {
    const { user } = yield auth.signInWithPopup(googleProvider);
    yield getSnapshotFromUserAuth(user);
  } catch (error) {
    yield put(signInFailure(error));
  }
}

export function* emailSignIn({ payload: { email, password } }) {
  try {
    const { user } = yield auth.signInWithEmailAndPassword(email, password);
    yield getSnapshotFromUserAuth(user);
  } catch (error) {
    yield put(signInFailure(error));
  }
}

export function* userSession() {
  try {
    const userAuth = yield getCurrentUser();
    if (!userAuth) return;
    yield getSnapshotFromUserAuth(userAuth);
  } catch (error) {
    yield put(signInFailure(error));
  }
}

export function* signOut() {
  try {
    yield auth.signOut();
    yield put(signOutSuccess());
  } catch (error) {
    yield put(signOutFailure(error));
  }
}

export function* signUp({ payload: { displayName, email, password } }) {
  try {
    const { user } = yield auth.createUserWithEmailAndPassword(email, password);
    yield put(signUpSuccess({ user, additionalData: { displayName } }));
  } catch (error) {
    yield put(signUpFailure(error));
  }
}

export function* signInOnSignUp({ payload: { user, additionalData } }) {
  yield getSnapshotFromUserAuth(user, additionalData);
}

export function* onStartSignUp() {
  yield takeLatest(UserActionTypes.START_SIGN_UP, signUp);
}

export function* onSignUpSuccess() {
  yield takeLatest(UserActionTypes.SIGN_UP_SUCCESS, signInOnSignUp);
}

export function* onStartGoogleSignIn() {
  yield takeLatest(UserActionTypes.START_GOOGLE_SIGN_IN, googleSignIn);
}

export function* onStartEmailSignIn() {
  yield takeLatest(UserActionTypes.START_EMAIL_SIGN_IN, emailSignIn);
}

export function* onCheckUserSession() {
  yield takeLatest(UserActionTypes.CHECK_USER_SESSION, userSession);
}

export function* onStartSignOut() {
  yield takeLatest(UserActionTypes.START_SIGN_OUT, signOut);
}

export function* userSagas() {
  yield all([
    call(onStartGoogleSignIn),
    call(onStartEmailSignIn),
    call(onCheckUserSession),
    call(onStartSignOut),
    call(onStartSignUp),
    call(onSignUpSuccess)
  ]);
}
