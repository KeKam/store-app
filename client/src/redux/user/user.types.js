const UserActionTypes = {
  SET_CURRENT_USER: 'SET_CURRENT_USER',
  START_GOOGLE_SIGN_IN: 'START_GOOGLE_SIGN_IN',
  START_EMAIL_SIGN_IN: 'START_EMAIL_SIGN_IN',
  SIGN_IN_SUCCESS: 'SIGN_IN_SUCCESS',
  SIGN_IN_FAILURE: 'SIGN_IN_FAILURE',
  CHECK_USER_SESSION: 'CHECK_USER_SESSION',
  START_SIGN_OUT: 'START_SIGN_OUT',
  SIGN_OUT_SUCCESS: 'SIGN_OUT_SUCCESS',
  SIGN_OUT_FAILURE: 'SIGN_OUT_FAILURE',
  START_SIGN_UP: 'START_SIGN_UP',
  SIGN_UP_SUCCESS: 'SIGN_UP_SUCCESS',
  SIGN_UP_FAILURE: 'SIGN_UP_FAILURE'
};

export default UserActionTypes;
