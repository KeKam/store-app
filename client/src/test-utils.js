import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';

import rootReducer from './redux/root-reducer';

// const render = (
//   component,
//   {
//     initialState,
//     store = createStore(rootReducer, initialState),
//     ...renderOptions
//   } = {}
// ) => {
//   return {
//     ...rtlRender(<Provider store={store}>{component}</Provider>),
//     store,
//   };
//   // const Wrapper = ({ children }) => {
//   //   return <Provider store={store}>{children}</Provider>;
//   // };
//   // return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
// };

const render = (component, { initialState } = {}) => {
  const actions = [];
  const observerMiddleware = () => (next) => (action) => {
    actions.push(action);
    return next(action);
  };
  const store = createStore(
    rootReducer,
    initialState,
    applyMiddleware(observerMiddleware)
  );
  const utils = {
    dispatch(action) {
      return store.dispatch(action);
    },
    getDispatchedActions() {
      return actions;
    },
    getState() {
      return store.getState();
    },
  };
  return {
    ...rtlRender(<Provider store={store}>{component}</Provider>),
    ...utils,
  };
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { render };
