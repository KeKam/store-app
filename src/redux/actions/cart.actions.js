import { CartActionTypes } from '../types/cart.types';

export const toggleCart = () => {
  return {
    type: CartActionTypes.TOGGLE_CART
  };
};

export const addToCart = item => {
  return {
    type: CartActionTypes.ADD_TO_CART,
    payload: item
  };
};

export const removeFromCart = item => {
  return {
    type: CartActionTypes.REMOVE_FROM_CART,
    payload: item
  };
};

export const removeItem = item => {
  return {
    type: CartActionTypes.REMOVE_ITEM,
    payload: item
  };
};
