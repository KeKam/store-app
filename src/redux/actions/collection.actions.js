import { CollectionActionTypes } from '../types/collection.types';
import {
  firestore,
  convertCollectionSnapshotToMap
} from '../../firebase/firebase';

export const startFetchCollection = () => ({
  type: CollectionActionTypes.START_FETCH_COLLECTION
});

export const fetchCollectionSuccess = collectionMap => ({
  type: CollectionActionTypes.FETCH_COLLECTION_SUCCESS,
  payload: collectionMap
});

export const fetchCollectionFailure = error => ({
  type: CollectionActionTypes.FETCH_COLLECTION_FAILURE,
  payload: error
});

export const startAsyncFetchCollection = () => {
  return async dispatch => {
    const collectionRef = firestore.collection('collection');
    dispatch(startFetchCollection());

    try {
      const response = await collectionRef.get();
      const collectionMap = convertCollectionSnapshotToMap(response);
      dispatch(fetchCollectionSuccess(collectionMap));
    } catch (error) {
      dispatch(fetchCollectionFailure(error));
    }
  };
};
