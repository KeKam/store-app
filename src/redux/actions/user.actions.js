export const SET_USER = user => {
  return {
    type: 'SET_USER',
    payload: user
  };
};
