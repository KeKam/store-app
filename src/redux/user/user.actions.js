import UserActionTypes from './user.types';

export const startGoogleSignIn = () => ({
  type: UserActionTypes.START_GOOGLE_SIGN_IN
});

export const startEmailSignIn = emailAndPassword => ({
  type: UserActionTypes.START_EMAIL_SIGN_IN,
  payload: emailAndPassword
});

export const signInSuccess = user => ({
  type: UserActionTypes.SIGN_IN_SUCCESS,
  payload: user
});

export const signInFailure = error => ({
  type: UserActionTypes.SIGN_IN_FAILURE,
  payload: error
});
