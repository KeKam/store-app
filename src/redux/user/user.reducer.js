import UserActionTypes from './user.types';

const DEFAULT_STATE = {
  currentUser: null,
  error: null
};

const userReducer = (state = DEFAULT_STATE, action) => {
  switch (action.type) {
    case UserActionTypes.SIGN_IN_SUCCESS:
      return {
        ...state,
        currentUser: action.payload,
        error: null
      };
    case UserActionTypes.SIGN_IN_FAILURE:
      return {
        ...state,
        error: action.payload
      };
    default:
      return state;
  }
};

export default userReducer;
