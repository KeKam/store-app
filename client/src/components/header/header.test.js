import React from 'react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';

import { render, screen } from '../../test-utils';
import { startSignOut } from '../../redux/user/user.actions';
import Header from './header';

const history = createMemoryHistory();

describe('Header component', () => {
  let mockState;

  beforeEach(() => {
    mockState = {
      currentUser: {
        email: 'testUser@gmail.com',
        password: 'testUser123',
      },
    };

    history.push = jest.fn();
  });

  it('should render Header component', () => {
    const { asFragment } = render(
      <Router history={history}>
        <Header />
      </Router>
    );

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByRole('link', { name: '4 SEASONS' })).toBeInTheDocument();
    expect(
      screen.getByRole('link', { name: 'COLLECTION' })
    ).toBeInTheDocument();
    expect(screen.getByRole('link', { name: 'CONTACT' })).toBeInTheDocument();
    expect(screen.getByRole('link', { name: 'SIGN IN' })).toBeInTheDocument();
    expect(
      screen.queryByRole('link', { name: 'SIGN OUT' })
    ).not.toBeInTheDocument();
    expect(screen.queryByTestId('bag')).not.toBeInTheDocument();
  });

  it('should show sign out button and cart icon', () => {
    render(
      <Router history={history}>
        <Header />
      </Router>,
      { initialState: { user: mockState } }
    );

    expect(screen.getByRole('link', { name: 'SIGN OUT' })).toBeInTheDocument();
    expect(
      screen.queryByRole('link', { name: 'SIGN IN' })
    ).not.toBeInTheDocument();
    expect(screen.getByTestId('bag')).toBeInTheDocument();
    expect(screen.getByTestId('item-count')).toBeInTheDocument();
  });

  it('should history push on click', () => {
    render(
      <Router history={history}>
        <Header />
      </Router>
    );

    userEvent.click(screen.getByRole('link', { name: '4 SEASONS' }));

    expect(history.push).toHaveBeenCalledTimes(1);
    expect(history.push).toHaveBeenCalledWith('/');

    userEvent.click(screen.getByRole('link', { name: 'COLLECTION' }));

    expect(history.push).toHaveBeenCalledTimes(2);
    expect(history.push).toHaveBeenCalledWith('/collection');

    userEvent.click(screen.getByRole('link', { name: 'CONTACT' }));

    expect(history.push).toHaveBeenCalledTimes(3);
    expect(history.push).toHaveBeenCalledWith('/contact');

    userEvent.click(screen.getByRole('link', { name: 'SIGN IN' }));

    expect(history.push).toHaveBeenCalledTimes(4);
    expect(history.push).toHaveBeenCalledWith('/login');
  });

  it('should dispatch startSignOut action and history push on click', () => {
    const { getDispatchedActions } = render(
      <Router history={history}>
        <Header />
      </Router>,
      { initialState: { user: mockState } }
    );

    userEvent.click(screen.getByRole('link', { name: 'SIGN OUT' }));

    expect(getDispatchedActions()).toContainEqual(startSignOut());
    expect(history.push).toHaveBeenCalledTimes(1);
    expect(history.push).toHaveBeenCalledWith('/');
  });

  it('should show CartDropdown component', () => {
    render(
      <Router history={history}>
        <Header />
      </Router>,
      { initialState: { user: mockState } }
    );

    userEvent.click(screen.getByTestId('bag'));

    expect(screen.getByText('You have no added items')).toBeInTheDocument();
  });
});
