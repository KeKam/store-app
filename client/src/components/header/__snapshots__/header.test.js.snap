// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should render Header component 1`] = `
<DocumentFragment>
  <div
    class="sc-jTzLTM jTGbTz"
  >
    <div
      class="sc-fjdhpX cFtmev"
    >
      <a
        class="sc-kAzzGY qPVRJ"
        href="/"
      >
        <div
          class="sc-kgoBCf dvjZwE"
        >
          4
        </div>
        <div
          class="sc-chPdSV cgZKXE"
        >
          SEASONS
        </div>
      </a>
    </div>
    <div
      class="sc-jzJRlG gBHJeD"
    >
      <a
        class="sc-cSHVUG cNCaJc"
        href="/collection"
      >
        COLLECTION
      </a>
      <a
        class="sc-cSHVUG cNCaJc"
        href="/contact"
      >
        CONTACT
      </a>
      <a
        class="sc-cSHVUG cNCaJc"
        href="/login"
      >
        SIGN IN
      </a>
    </div>
  </div>
</DocumentFragment>
`;
